doc/step-in-groovy.md

```diff
@@ -0,0 +1,96 @@
+# Implementing Step in Groovy
+The standard way of implementing a step is by writing Java code that extends from `Step` and `StepDescriptor`,
+but you can also define a step in Jenkins Pipeline Script.
+
+This approach is quite convenient when your step composes other
+existing steps. For example, the following files define
+the equivalent of the `retry` step that repeats the body up to several times:
+
+```java
+// src/main/java/acme/RetryStep.java
+package acme;
+
+class RetryStep extends GroovyStep {
+    private int times;
+
+    @DataBoundConstructor
+    public RetryStep(int times) {
+        this.times = times;
+    }
+
+    public int getTimes() {
+        return times;
+    }
+
+    @Extension
+    public static class DescriptorImpl extends GroovyStepDescriptor {
+        @Override public String getFunctionName() {
+	        return "myRetryClone";
+        }
+
+        @Override public String getDisplayName() {
+	        return "Retry step";
+        }
+    }
+}
+```
+
+```groovy
+// src/main/resources/acme/RetryStepExecution.groovy
+package acme;
+
+public class RetryStepExecution extends GroovyStepExecution {
```

Comment:
Okay.
I'm from the camp that the point of an example is to highlight important bits, and if compiler accuracy gets in the way it should be omitted.